### PR TITLE
textproto: check header characters

### LINKED ISSUE #80: RFC 5322 character use limitation
RFC 5322 limits use of characters in general (US-ASCII) and for some specific scopes. For example in headers only printable ASCII characters (`'!'..'~'`), SPACE and HTAB are allowed and `\r\n ` sequence is used for line folding.

It's currently possible to pass some disallowed symbols to go-message. The most noticeably it's a single `\n` in header values which is treated as a regular symbol or a part of `\r\n ` sequence. As it was discussed in #77, the best way of handling such cases is just failing with an explicit error.

I have a few possible solutions. but I don't feel like any of them is clearly "the best":
* checking whether `\n` is a part of the folding sequence; solves only a subproblem, but does not change behavior on externally folded line
* Matching strings with a regex (`"[\t -~]*"` for headers); probably the most clean, but may be not great in terms of performance
* Manually checking strings symbol by symbol, may introduce some possible optimizations like single pass `foldLine` but certainly is not as fancy as the previous one.

I don't know what would be better to implement (or maybe I don't see something else), so I'd like to discuss it here before coding.

### PR DESCRIPTION
**Pull request by dvalter**
Add checks for header field name according to RFC 6532 and disallow newline characters in field values.

Resolves #80

### CODE REVIEW

File: textproto/header.go [hunk 1]
```diff
@@ -87,9 +98,20 @@ func (h *Header) AddRaw(kv []byte) {
 	h.m[k] = append(h.m[k], f)
 }
 
+var headerNameRegex = regexp.MustCompile("^[!-9;-~]*$")
```
Comment (foxcpp) [inline textproto/header.go:101]: That was a bit confusing to read at first. Perhaps a little commentary can help (e.g. "matches string of ASCII printable characters except for colon") or [just get rid of the regular expression here](https://xkcd.com/1171/) and replace it with a loop.

**Comment by foxcpp** [comment]
LGTM otherwise.

File: textproto/header.go (hunk 1 above)
Comment (dvalter) [reply textproto/header.go]: You're 100% right here.
I hope now `':'` is highlighted more explicitly now.

**Comment by emersion** [comment]
Question: should `Header.Add` return an error, or should we return an error when formatting the header?

- Returning an error in `Header.Add` makes it a little bit annoying for callers to add headers since they need to error-check each time. Also, existing users may have header fields silently dropped, so this may be confusing/annoying to debug.
- Returning an error when formatting the header may make it more difficult to trace which `Header.Add` call is responsible for the malformed header.

**Comment by foxcpp** [comment]
>Returning an error when formatting the header may make it more difficult to trace which Header.Add call is responsible for the malformed header.

Adding enough information to the error message to identify exact field that caused it can help with it.

I agree with your concerns regarding Header.Add. It is unexpected that Header.Add can fail and some people just ignore errors if it gets too verbose and "cannot possibly fail here".

**Comment by foxcpp** [comment]
Therefore I think it is a good idea to move check to the formatting time and make error look like "Nth DKIM-Signature value contains invalid characters"

**Comment by dvalter** [comment]
I moved checks to `WriteHeader`. However I'm not sure about error message format and the fact that we go there from the last header towards the first and error in header `k` does not mean any of `k-n` is fine. 

Regarding rebase to master, in c5125629c header type check was moved to a separate function, and I honestly have no idea how keep it clean and informative about what exact header is incorrect

**Comment by emersion** [comment]
Maybe this?

```go
if _, err := w.Write(f.raw()); err != nil {
    return fmt.Errorf("failed to write header field #%v (%q): %w", i, f.k, err)
}
```

**Comment by emersion** [comment]
But I wonder if it's a good idea to export `DisallowedCharacterError` at all. Maybe we should wait for someone to ask for it. Depending on the use-case it may make sense to have e.g. `HeaderFieldError`.

**Comment by dvalter** [comment]
> Maybe this?

In your example `err` is returned by writer.
There is the closest code I can come up with is this, in case we add this check to `raw()` or something close if there will be another validation function.
```go
if rawField, err := f.raw(); err == nil {
	if _, err := w.Write(rawField); err != nil {
		return err
	}
} else {
	return fmt.Errorf("failed to write header field #%v (%q): %w", len(h.l) - i, f.k, err)
}
```

**Pull request merged**